lib/puppet/pops/types/p_encrypted_type.rb

```diff
@@ -0,0 +1,297 @@
+#
+# @api public
+class PEncryptedType < PAnyType
+
+  # Represents an Encrypted value
+  # @api public
+  #
+  class Encrypted
+    attr_reader :format
+    attr_reader :encrypted_key
+    attr_reader :crypt
+    attr_reader :encrypted_fingerprint
+
+    # Creates a new Encrypted
+    # This API is private - please use `Encrypted.encrypt` method to create an instance.
+    # This method is only accessible for serialization purposes.
+    #
+    # @api private
+    #
+    def initialize(format, encrypted_key, crypt, encrypted_fingerprint)
+      binary = Puppet::Pops::Types::PBinaryType::Binary
+      @format = format
+      @encrypted_key, @crypt, @encrypted_fingerprint = make_binary(encrypted_key, crypt, encrypted_fingerprint)
+    end
+
+    # @api private
+    def make_binary(*vars)
+      vars.map {|v| Puppet::Pops::Types::PBinaryType::Binary.from_binary(v) }
+    end
+    private :make_binary
+
+    # Presents the binary content as a base64 encoded string (without line breaks).
+    # While this method is public, the content of the String is not API and may change.
+    #
+    # @api public
+    #
+    def to_s
+      [ @format,
+        @encrypted_key,
+        @crypt,
+        @encrypted_fingerprint,
+        ].to_s
+    end
+
+    # @api public
+    #
+    def hash
+      # this is the encrypted random key which should give a reasonable spread
+      @encrypted_key.hash
+    end
+
+    # Returns identity equality since two Encrypted for the same value will always be different.
+    # @param o [Object] Any object to check for equality
+    # @return [Boolean]
+    # @api public
+    #
+    def eql?(o)
+      # Can only equal the same instance since each encryption is based on random values
+      # No need to check for anything besides object identity
+      __id__ == o.__id__
+    end
+
+    # Returns identity equality since two Encrypted for the same value will always be different.
+    # @param o [Object] Any object to check for equality
+    # @return [Boolean]
+    # @api public
+    #
+    def ==(o)
+      self.eql?(o)
+    end
+
+    # Encrypts data using a certificate, a cipher name and Any data - returns an Encrypted.
+    # This is the API for creating an `Encrypted` value.
+    #
+    # @param certificate [Puppet::SLL::Certificate] The certificate to use (i.e. for the receiver's public key)
+    # @param cipher [String] The name of the cipher to use - for example 'AES-256-CBC'
+    # @param data [Object] Any puppet rich data data type
+    # @return [Puppet::Pops::Types::PEncryptedType::Encrypted] An Encrypted puppet data type value
+    # @api public
+    #
+    def self.encrypt(certificate, cipher, data)
+      key = certificate.content.public_key
+      fingerprint = certificate.fingerprint
+      format = 'json,' + cipher
+      data = serialize(data)
+
+      aes_encrypt = OpenSSL::Cipher.new(cipher).encrypt
+
+      # Use a random key
+      aes_encrypt.key = aes_key = aes_encrypt.random_key
```

Comment:
I get that this is safer but you might want to have an option where a global key is generated for a given catalog ID to take some crypto load off of the server.

It will also take load off of the client since it would know that it's a global value and could then do everything else symmetrically.

I understand that this makes the data in the catalog more susceptible to offline brute forcing, but I think that the user should be able to decide on that tradeoff with the more secure version being the default.

Reply:
I understand that as generating the random key once which would not be hard on the server, on the client side is is more problematic as I have no way of sending the random key just once (at least not without modifying the catalog format).

Reply:
@hlindberg You couldn't just drop in a `key` resource or something and encrypt the payload there? Or, without creating a new resource, a `notify` resource at `loglevel` debug that just says "don't blink" or something like that?

Reply:
@hlindberg In terms of the modifying the catalog format, it's just serialized, correct? If that's the case, wouldn't this just be a top level object added to the catalog for Puppet 6.?.?+.

Also, technically you could just encrypt the entire catalog with the target client's public key prior to sending which would solve the "at rest on the client" problem as well but may be VERY expensive.

Reply:
Full catalog compilation at REST is already available via some module. 
So thousands of encrypt in a catalog - that would benefit from the simple optimization of using the same random key for the entire catalog even if serialized multiple times.
On the agent it would take the same amount of time, but that is a somewhat lesser concern.

Regarding catalog; yes it is "just" adding a top level field, but that affects downstream projects (octocatalog-diff, catalog-preview, comes to mind). Those tools probably have to update anyway come to think of it because the encrypted values will always be different due to being encrypted with a random key. Diffs are not really meaningful.

Reply:
+1 would be nice to generate one AES key per compile, not one key per secret.